File: elixir/2270_ways_to_split_array.exs
# https://leetcode.com/problems/number-of-ways-to-split-array/
# 03/01/2025

defmodule Solution do
  @spec ways_to_split_array(nums :: [integer]) :: integer
  def ways_to_split_array(nums) do
    sum = nums |> Enum.sum()

    nums
    |> Enum.take(length(nums) - 1)
    |> Enum.reduce({0, 0}, fn num, {count, left_sum} ->
      new_left_sum = left_sum + num
      new_count = if new_left_sum >= sum - new_left_sum, do: count + 1, else: count
      {new_count, new_left_sum}
    end)
    |> Kernel.elem(0)
  end
end

defmodule Main do
  def main do
    test_cases = [
      %{
        nums: [10, 4, -8, 7],
        expected: 2
      },
      %{
        nums: [2, 3, 1, 0],
        expected: 2
      }
    ]

    Enum.each(test_cases, fn %{nums: nums, expected: expected} ->
      result = Solution.ways_to_split_array(nums)
      IO.puts("Input: nums=#{inspect(nums)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/2270_ways_to_split_array.rs
// https://leetcode.com/problems/number-of-ways-to-split-array/
// 03/01/2025

impl Solution {
    pub fn ways_to_split_array(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let prefix_sum: Vec<i64> = nums
            .iter()
            .scan(0, |state, &num| {
                *state += num as i64;
                Some(*state)
            })
            .collect();
        let sum = prefix_sum[n - 1];

        prefix_sum
            .iter()
            .take(n - 1)
            .fold(0, |acc, &num| if num >= sum - num { acc + 1 } else { acc })
    }
}


#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case(
        [10,4,-8,7].to_vec(),
        2
    )]
    #[case(
        [2,3,1,0].to_vec(),
        2
    )]
    #[case(
        [6, -1, 9].to_vec(), 
        0
    )]
    fn test_vowel_strings(#[case] nums: Vec<i32>, #[case] expected: i32) {
        let result = Solution::ways_to_split_array(nums);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        ([10, 4, -8, 7].to_vec(), 2),
        ([2, 3, 1, 0].to_vec(), 2),
        ([6, -1, 9].to_vec(), 0),
    ];

    for (nums, expected) in test_cases {
        let result = Solution::ways_to_split_array(nums.clone());
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            nums, expected, result
        );
        assert_eq!(result, expected);
    }
}
